group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,228 @@
+                // 3) If remaining > 0: member doesn't have the minimum required partitions, so it should be added to potentiallyUnfilledMembers.
+                int remaining = minRequiredQuota - currentAssignmentSize;
+
+                // In order to remain as sticky as possible, since the order of members can be different, we want the members that have more than the required number of partitions,
+                // to get the extra partition instead of assigning the extras to the first few members directly.
+                if (remaining < 0 && numMembersWithExtraPartition > 0) {
+                    numMembersWithExtraPartition--;
+                    // Since we already added the minimumRequiredQuota of partitions in the previous step (until minReq - 1), we just need to
+                    // add the extra partition that will be present at the index right after min quota was satisfied.
+                    assignedStickyPartitionsForTopic
+                        .add(currentAssignmentListForTopic.get(minRequiredQuota));
+                    newAssignment.computeIfAbsent(memberId, k -> new MemberAssignment(new HashMap<>()))
+                        .targetPartitions()
+                        .computeIfAbsent(topicId, k -> new HashSet<>())
+                        .add(currentAssignmentListForTopic.get(minRequiredQuota));
+                } else {
+                    MemberWithRemainingAssignments newPair = new MemberWithRemainingAssignments(memberId, remaining);
+                    potentiallyUnfilledMembers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // Find the difference between the total partitions per topic and the already assigned sticky partitions for the topic to get unassigned partitions.
+            // List of unassigned partitions per topic contains the partitions in ascending order.
+            List<Integer> unassignedPartitionsForTopic = new ArrayList<>();
+            for (int i = 0; i < numPartitionsForTopic; i++) {
+                if (!assignedStickyPartitionsForTopic.contains(i)) {
+                    unassignedPartitionsForTopic.add(i);
+                }
+            }
+
+            // Step 4
+            // If remaining > 0 after increasing the required quota due to the extra partition, add potentially unfilled member to the unfilled members list.
+            int unassignedPartitionsListStartPointer = 0;
+            for (MemberWithRemainingAssignments pair : potentiallyUnfilledMembers) {
+                String memberId = pair.memberId;
+                Integer remaining = pair.remaining;
+                if (numMembersWithExtraPartition > 0) {
+                    remaining++;
+                    numMembersWithExtraPartition--;
+                }
+                if (remaining > 0) {
+                    // Step 5
+                    // Assign the remaining number of partitions from the unassigned partitions list.
```

Comment:
Let's remove step 5 here and include it in step 4.

Reply:
I think its fine honestly, its different from step 4. If its too much in one step there isn't really much point in breaking it up right?

Reply:
okayyy